sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/ScalaUdf.scala

```diff
@@ -49,353 +49,378 @@ case class ScalaUdf(function: AnyRef, dataType: DataType, children: Seq[Expressi
+            ScalaReflection.convertToScala(children(14).eval(input), children(14).dataType),
+            ScalaReflection.convertToScala(children(15).eval(input), children(15).dataType),
+            ScalaReflection.convertToScala(children(16).eval(input), children(16).dataType),
+            ScalaReflection.convertToScala(children(17).eval(input), children(17).dataType),
+            ScalaReflection.convertToScala(children(18).eval(input), children(18).dataType),
+            ScalaReflection.convertToScala(children(19).eval(input), children(19).dataType),
+            ScalaReflection.convertToScala(children(20).eval(input), children(20).dataType))
+        }
+      case 22 => 
+        val func = function.asInstanceOf[(Any, Any, Any, Any, Any, Any, Any, Any, Any, Any, Any, Any, Any, Any, Any, Any, Any, Any, Any, Any, Any, Any) => Any]
+        (input: Row) => {
+          func(
+            ScalaReflection.convertToScala(children(0).eval(input), children(0).dataType),
+            ScalaReflection.convertToScala(children(1).eval(input), children(1).dataType),
+            ScalaReflection.convertToScala(children(2).eval(input), children(2).dataType),
+            ScalaReflection.convertToScala(children(3).eval(input), children(3).dataType),
+            ScalaReflection.convertToScala(children(4).eval(input), children(4).dataType),
+            ScalaReflection.convertToScala(children(5).eval(input), children(5).dataType),
+            ScalaReflection.convertToScala(children(6).eval(input), children(6).dataType),
+            ScalaReflection.convertToScala(children(7).eval(input), children(7).dataType),
+            ScalaReflection.convertToScala(children(8).eval(input), children(8).dataType),
+            ScalaReflection.convertToScala(children(9).eval(input), children(9).dataType),
+            ScalaReflection.convertToScala(children(10).eval(input), children(10).dataType),
+            ScalaReflection.convertToScala(children(11).eval(input), children(11).dataType),
+            ScalaReflection.convertToScala(children(12).eval(input), children(12).dataType),
+            ScalaReflection.convertToScala(children(13).eval(input), children(13).dataType),
+            ScalaReflection.convertToScala(children(14).eval(input), children(14).dataType),
+            ScalaReflection.convertToScala(children(15).eval(input), children(15).dataType),
+            ScalaReflection.convertToScala(children(16).eval(input), children(16).dataType),
+            ScalaReflection.convertToScala(children(17).eval(input), children(17).dataType),
+            ScalaReflection.convertToScala(children(18).eval(input), children(18).dataType),
+            ScalaReflection.convertToScala(children(19).eval(input), children(19).dataType),
+            ScalaReflection.convertToScala(children(20).eval(input), children(20).dataType),
+            ScalaReflection.convertToScala(children(21).eval(input), children(21).dataType))
```

Comment:
`children` is the type of `Seq[Expression]` (essentially the `List[Expression]`), access its element by index will cause performance overhead in runtime, we'd better move that out of the anonymous functions. See:
http://docs.scala-lang.org/overviews/collections/performance-characteristics.html